posts/2021/07/moving-scipy-to-meson.md

```diff
@@ -0,0 +1,489 @@
+to debug, but in some cases it can restrict what you can do. For example,
+this code for building all `scipy.sparse.csgraph` extensions is quite
+elegant, but the `foreach` pattern will not work if the `.pyx` files are
+generated. This is because they'd then be objects created by `custom_target`,
+and there is no syntax to give them unique names within the `foreach` loop -
+this cost me some time to figure out:
+```python
+pyx_files = [
+  ['_flow', '_flow.pyx'],
+  ['_matching', '_matching.pyx'],
+  ['_min_spanning_tree', '_min_spanning_tree.pyx'],
+  ['_reordering', '_reordering.pyx'],
+  ['_shortest_path', '_shortest_path.pyx'],
+  ['_tools', '_tools.pyx'],
+  ['_traversal', '_traversal.pyx'],
+]
+
+foreach pyx_file : pyx_files
+  py3.extension_module(pyx_file[0],
+    pyx_file[1],
+    include_directories : inc_np,
+    dependencies : py3_dep,
+    install : true,
+    subdir : 'scipy/sparse/csgraph')
+endforeach
+```
+
+*Meson uses `pkg-config` for discovering dependencies.* This means certain
+things "just work". E.g., `blas = dependency('openblas')` followed by using
+`dependencies: blas` to build the `scipy.linalg._fblas` extension worked on
+the first try for me - definitely did not expect that.
+
+*There is one escape hatch to do things Meson does not natively support: `custom_target`.*
+As long you list its source files and outputs, you can invoke Python scripts
+through `custom_target`. This is how the SciPy build invokes `numpy.f2py`,
+Pythran and `cython --cplus` (Meson's Cython support is brand new, and only
+supports targeting C if you give a `.pyx` file to `py3.extension_module`
+right now).
+
+
+## The sharp bits
+
+Not everything was smooth sailing. Here are a few of the most important
+issues I ran into:
+
+1. Cython + generated sources. This was by far the most time-consuming topic.
+   Cython's `cimport` mechanism relies on the complete source tree. For
+   example, not having a `__init__.py` file two directories up from where the
+   `.pyx` file that you are building lives changes the C code that Cython
+   generates. The full source tree layout matters for Cython. Cython was
+   largely designed around in-tree builds, and Meson only allows you to
+   generate files out-of-tree. This necessitates hacks like writing scripts
+   or invoking `cp` to copy over many `__init__.py` and `.pxi|.pxd` files.
+   And then more hacks to ensure those files are respected as dependencies by
+   Ninja.
+2. SciPy is a tangled mess - there are 17 `scipy.xxx` submodules and almost all
+   of them depend on each other. So with a few exceptions, the tests only ran
+   once almost all submodules built with Meson. We have had issues in SciPy
+   with import cycles before, and I'm now surprised we didn't have more
+   issues in the past.
+3. One feature I missed in Meson: installing generated files with
+   `py3.install_sources` is not allowed. One can specify `install: true` in
```

Comment:
link to `py3.install_sources`? Are there ideas how to mitigate this restriction moving forward?

Reply:
added a link. yes, we can just implement this in Meson, it looks like it's something that not many people needed so far.